dbms/src/Functions/greatCircleDistance.cpp

```diff
@@ -103,16 +184,30 @@ class FunctionGreatCircleDistance : public IFunction
             lat1Deg < -90 || lat1Deg > 90 ||
             lat2Deg < -90 || lat2Deg > 90)
         {
-            throw Exception("Arguments values out of bounds for function " + getName(), ErrorCodes::ARGUMENT_OUT_OF_BOUND);
+            throw Exception("Arguments values out of bounds for function " + getName(),
+                            ErrorCodes::ARGUMENT_OUT_OF_BOUND);
         }
 
-        Float64 lon1Rad = degToRad(lon1Deg);
-        Float64 lat1Rad = degToRad(lat1Deg);
-        Float64 lon2Rad = degToRad(lon2Deg);
-        Float64 lat2Rad = degToRad(lat2Deg);
-        Float64 u = sin((lat2Rad - lat1Rad) / 2);
-        Float64 v = sin((lon2Rad - lon1Rad) / 2);
-        return 2.0 * EARTH_RADIUS_IN_METERS * asin(sqrt(u * u + cos(lat1Rad) * cos(lat2Rad) * v * v));
+        float dlat = GeodistDegDiff(lat1Deg - lat2Deg);
+        float dlon = GeodistDegDiff(lon1Deg - lon2Deg);
+
+        if (dlon < 13)
+        {
+            // points are close enough; use flat ellipsoid model
+            // interpolate sqr(k1), sqr(k2) coefficients using latitudes midpoint
+            float m = (lat1Deg + lat2Deg + 180) * GEODIST_TABLE_K / 360; // [-90, 90] degrees -> [0, KTABLE] indexes
+            int i = static_cast<int>(m);
+            i &= (GEODIST_TABLE_K - 1);
+            float kk1 = g_GeoFlatK[i][0] + (g_GeoFlatK[i + 1][0] - g_GeoFlatK[i][0]) * (m - i);
```

Comment:
If you have some explanation on algorithm, please provide it here (and rename variables accordingly).